# የስብጥርነት መለኪያዎች

የስብጥርነት ወይም የመበታተንነት መለኪያዎች በውሂባችን ውስጥ የሚኖረውን የእሴት ልዩነት ለመለካት የሚያስችሉን ናቸው። ይህንን ስብጥር የምናጠቃልልበት መለኪያዎቻችን የስብጥርነት መለኪያዎች ይባላሉ። የመጀመሪያ የምንመለከተው አካላይነት ወይም ሬንጅ የምንለውን ይሆናል። በመቀጠልም ስታንዳርድ ዲቪየሽንን እና ልዩነት (ቫሪያንስን) እንመለከታለን። 

*** 
ለዜሮ የቀረበ የስብጥርነት እሴት ውሂቡ የተቀራረበ መሆኑን ያሳያል። የስብጥርነት መለኪያው እየጨመረ ሲሄድ ውሂቡ የተሰበጣጠረ መሆኑን ይጠቁማል። 
*** 

## ማዕከላዊነት ክለሳ

በማእከላዊነት ያየናቸውን መለኪያዎች (አማካኝ፣ አጋማሽ እና ተደጋጋሚው እሴት) ተግባር እንጻፍ እና ድጋሚ እንቀምራቸው። 

In [1]:
# ፋይሉን ማምበብ 
import csv
with open("./ውሂብ/የምረቃ_ውሂብ.csv", 'r', encoding='utf-8') as ፋይል:
    ውሂብ = list(csv.reader(ፋይል))
    ዲፕሎማ_ተመራቂ = [float(ዲ[2]) for ዲ in ውሂብ[1:]] 
def አማካኝ(ቁጥር):
    """ አማካኝ የሚቀምር ተግባር"""
    return (sum(ቁጥር)/len(ቁጥር))

def አጋማሽ(ቁጥር):
    """ አጋማሽ የሚቀመር ተግባር """
    ረድፍ_የያዘ_ቁጥር = sorted(ቁጥር)
    ብዛት = len(ቁጥር)
    አጋማሹ_ቦታ = ብዛት//2 # ኢንቲጀር ማካፈል ነው 
    # ብዛቱ ጎዶሎ ቁጥር ከሆነ 
    if ብዛት%2==1:
        return ረድፍ_የያዘ_ቁጥር[አጋማሹ_ቦታ] 
    else:
        # ብዛቱ ሙሉ ቁጥር ከሆነ 
        ከላይ_ያለው = አጋማሹ_ቦታ
        ከታች_ያለው = አጋማሹ_ቦታ - 1
        return (ረድፍ_የያዘ_ቁጥር[ከላይ_ያለው] + ረድፍ_የያዘ_ቁጥር[ከታች_ያለው])/2    
    
def ተደጋጋሚው_እሴት(ቁጥር):
    """ ተደጋጋሚውን እሴት የሚቀምር ተግባር"""
    ተደጋጋሚ_ቆጠራ = {}
    ዲፕሎማ_ተመራቂ = [float(ዲ[2]) for ዲ in ውሂብ[1:]] 
    for እሴት in ዲፕሎማ_ተመራቂ:
        if እሴት not in ተደጋጋሚ_ቆጠራ:
            ተደጋጋሚ_ቆጠራ[እሴት] =1
        else:
            ተደጋጋሚ_ቆጠራ[እሴት]+=1
        
    # ተደጋጋሚውን እሴት ለማግኘት 
    # ተደጋጋሚው ዜሮ ግዜ ነው የተደጋገመው ብለን ታሳቢ አድርገን እንነሳ
    የተደጋጋሚው_ቆጠራ_ብዛት = 0
    ተደጋጋሚው_እሴት = None

    for ቁልፍ, እሴት in ተደጋጋሚ_ቆጠራ.items():
        if የተደጋጋሚው_ቆጠራ_ብዛት < እሴት:
            የተደጋጋሚው_ቆጠራ_ብዛት = እሴት
            ተደጋጋሚው_እሴት=ቁልፍ
    return(ተደጋጋሚው_እሴት) #, የተደጋጋሚው_ቆጠራ_ብዛት)   
def አነስተኛ (ቁጥር):
    """ ትንሹን ቁጥር የሚመልስ"""
    return min(ቁጥር)

def ከፍተኛ(ቁጥር):
    """ከፍተኛውን ቁጥር የሚመልስ"""
    return max(ቁጥር)

In [2]:
# አማካኝ 
አማካኝ(ዲፕሎማ_ተመራቂ)

770.7796610169491

In [13]:
# አማካኝ 
አጋማሽ(ዲፕሎማ_ተመራቂ)

679.0

In [14]:
ተደጋጋሚው_እሴት(ዲፕሎማ_ተመራቂ)

1.0

In [16]:
አነስተኛ(ዲፕሎማ_ተመራቂ)

1.0

In [17]:
ከፍተኛ(ዲፕሎማ_ተመራቂ)

3299.0

## አካላይነት (ሬንጅ)

አካላይነት ወይም ሬንጅ የምንለው የስብጥርነት መለኪያ ከሆኑት አንዱ ነው። ዝቅተኛውን ከከፍተኛው በማቀናነስ የሚገኝ እሴት ነው። 

In [18]:
# አካላይነትን ለመቀመር 
def አካላይነት (ቁጥር):
    """አካላይነትን የሚሰራ ቁጥር"""
    return (max(ቁጥር) - min(ቁጥር))

In [19]:
# የዲፕሎማ ተመራቂ ተማሪዎችን አካላይነት ለመስራት ስለሆነ የምንፈልገው
# የዲፕሎማ ተመራቂዎችን ዝርዝር እንውሰድ 
ዲፕሎማ_ተመራቂ = [float(ዲ[2]) for ዲ in ውሂብ[1:]]
አካላይነት(ዲፕሎማ_ተመራቂ)

3298.0

ይሔ ምን ማለት ነው? የውሂባችን አጋማሽ 679 ሲሆን አካላይነት ደግሞ 3298 ነው። በሁለቱ መካከል እጅግ ከፍተኛ የሆነ ልዩነት አለ። ይሄም ውሂባችን በስብጥርነት የተጠቃ መሆኑን ማረጋገጥ እንችላለን። ይሔንን ዋጋ ከዲግሪ ወይም ከድህረምረቃ ተመራዊ ተማሪዎች አካላይነት ጋር ማወዳደር እንችላለን። እንደዚህ አይነቱን ችግር ለመቅረፍ ኳንታይል እና ኢንተርኳርታይል አካላይነት የሚባሉ መለኪያዎች አሉ። እነሱም ለከፍተኛ ስብጥርነት ምክኒያት የሆኑትን ለይቶ ለማውጣት ይጠቅማሉ። 

### ኳንታይል

ምልከታችንን ወይም እሴታችንን በእኩል ይሆንታ ለመክፈል የሚረዳ ነው። ለምሳሌ የሚከተለው ውሂብ ቢኖረን እና በእኩል ይሆንታ ሁለት ሁለት እያደረግን ብንከፍል (20% ማለት ነው)፤ 80% ዊሂባችን የትኞቹን ምልከታዎች እንደሚይዝ መመልከት እንችላለን። ይህንንም ተጠቅመን ለስብጥርነት ምክኒያት የሆኑትን ከ ምልከታችን ቀንሰን ማውጣት እንችላለን።   `[55,44,22,18,96,44,77,87,62,83]` 

In [6]:
def ኳንታይል(እሴቶች, ፐርሰንታይል):
    """ከዝርዝር ውስጥ የተፈለገውን ፐርሰንታይል ቁጥር ይመልስልናል
    """
    
    ፐ_አመልካች = int(ፐርሰንታይል * len(እሴቶች))
    return sorted(እሴቶች)[ፐ_አመልካች]

### የ ዲፕሎማ_ተመራቂ ኳንታይልን ብንፈልግ (25% 50% 75% 90% እና 95%)
for ፐርሰንት in [0, 0.1, 0.25, 0.50, 0.75, 0.90, 0.95, 0.99]:
    ኳ = ኳንታይል(ዲፕሎማ_ተመራቂ, ፐርሰንት)
    
    print('ፐርሰንት: {}\t{}'.format(ፐርሰንት, ኳ))

ፐርሰንት: 0	1.0
ፐርሰንት: 0.1	15.0
ፐርሰንት: 0.25	92.0
ፐርሰንት: 0.5	679.0
ፐርሰንት: 0.75	1218.0
ፐርሰንት: 0.9	1726.0
ፐርሰንት: 0.95	1818.0
ፐርሰንት: 0.99	3299.0


ከዚህ ውጤት እንደምንረዳው 92 የ 25% ኳንታይል እሴት ነው። 1818ም የ 95% ኳንታይል እሴት ነው። ውሂባችንን በዚህ መንገድ ተመልክተን በ 25% እና በ 95% ኳንታይል ያሉትን ውሂቦች ብቻ ወስደን መተንተን እንችላለን ማለት ነው። 

### ኢንተር ኳንታይል አካላይነት

25% እና 75% ኳንታይ የተለመዱ እና ጥቅም ላይ የሚውሉ ኳንታይሎች ናቸው። ከላይ ለመጥቀስ እንደተሞከረው አላማው የማይፈለጉና በውሂባችን ውስጥ ከፍተኛ የሆነ ስብጥርነትን የሚፈጥሩትን ምልከታዎች ለመቀነስ ነው። እንደሚከተለው መቀመር እንችላለን። 

In [7]:
# ኢንተር ኳንታይል አካላይነትን ለመቀመር ጥቅም ላይ የዋለ ተግባር 
def ኢኳንታይል_አካላይነት(እሴቶች):
    '''በ 75% እና 25% ፐርሰንታይል መካከል ያለውን ልዩነት ይመልሳል።
    '''
    
    return ኳንታይል(እሴቶች, 0.75) - ኳንታይል(እሴቶች, 0.25)
# በዲፕሎማ_ተመራቂዎች ያለው የኢኳንታይል አካላይነትን ለማገኘት (በ 25% እና 75%)
ኢኳንታይል_አካላይነት(ዲፕሎማ_ተመራቂ)

1126.0

በዊሂባችን ያለውን ስብጥር ከ 3298 ወደ 1126 ማውረድ እንችላለን ማለት ነው ከ 25%(92) በታች እና 75%(1218) በላይ ያሉትን እሴቶች በመተው።   

In [23]:
# በ 25% እና 75% ብቻ ያለውን ከመተው፤ 
#በምንፈልገው ክልል ውስጥ ያለውን ለመተው የሚከተለውን መጠቀም እንችላለን።

def ኢኳንታይል_አካላይነት(እሴቶች, የላይኛው=0.75, የታችኛው=0.25):
    '''በ 75% እና 25% ፐርሰንታይል መካከል ያለውን ልዩነት ይመልሳል። በነባሪው ማለት ነው። 
    '''
    
    return ኳንታይል(እሴቶች, የላይኛው) - ኳንታይል(እሴቶች, የታችኛው)

ኢኳንታይል_አካላይነት(ዲፕሎማ_ተመራቂ, 0.9, 0.1) # በ 90% እና በ 10%

1711.0

የኳንታይልም ሆነ ኢንተር ኳርታይል አካላይነት የሚነግሩን በላይኛው እና በታችኛው ኳንታይል መካከል ያለውን ልዩነት ነው። እሴቶቹ ምን ያህል ግዜ እንደተደጋገሙ ወይም ወደ አማካኙ ምን ያህል እንደቀረቡ እና እንደራቁ አይነግሩንም። የእሴቶቹን ከአማካኝ ያላቸውን ርቀት ነጋሪ የሆኑ መለኪያዎችን እንደሚከተለው እንመለከታለን። 

## ልዩነት (ቫሪያንስ) 

ልዩነት (ቫሪያንስ) የምንለው እና ስታንዳርድ ዲቪየሽን የውሂባችንን ስብጥር በአማካኙ ዙርያ ምን ያክል እንደሆነ ይነግረናል። 

***
$\sigma^2 = \frac{\sum_{ኢ=1}^{የ} (አ{_i}-\bar{አ})^2 }{የ-1} \Rightarrow $ (ቁ)ጥር ከ አንድ እስከ (የ)ትዬለሌ ወይም በውሂባችን ውስጥ ያለውን ምልከታ ቁጥር ብዛት፤ (እ)ሴት የእያንዳንዱን ምልከታ ዋጋ ወይም እሴት ሲያመለክት $\bar{አ}$ አማካኝን ያመለክታል። 
***

In [22]:
# ለዩነትን ወይም ቫሪያንስን ለመቀመር 
def ልዩነት(ቁጥር):
    #ቅናሽን 0 አድርገን እንነሳ
    ቅናሽ = 0
    አማካኝ = sum(ቁጥር)/len(ቁጥር)
    #እያንዳንዱን እይታ አማካኙን እየቀነስን ስክዌር እናድርግ እና ቅናሽ ላይ እንደምረው 
    for እሴት in ቁጥር:
        ቅናሽ += (እሴት - አማካኝ)**(2)
    # ልዩነቱን በቁጥሩ ብዛት -1 አካፍሎ መመለስ 
    return (ቅናሽ)/(len(ቁጥር)-1)

In [23]:
# የዲፕሎማ ልዩነትን ለማግኘት 
ልዩነት(ዲፕሎማ_ተመራቂ)

486868.34716540034

ይኼንን ቁጥር ትርጉም ለመስጠት ስናስብ መለኪያው ከዊሂባችን መለኪያ ስክዌር የተደረገ መሆኑን መገንዘብ ያስፈልጋል። ማለትም በሜትር የተሰበሰብ ውሂብ ከሆነ ልዩነት የሚነግረን በሜትር ስክዌር ነው። ስለዚህም ተመሳሳይ በሆነ የመለኪያ ዩኒት የውሂባችንን ስብጥር የሚነግረን ስታንዳርድ ዲቪየሽን ነው። 

## ስታንዳርድ ዲቪየሽን

ስታንዳርድ ዲቪየሽን ልዩነትን በ ስክዌር ሩት ውስጥ ስናስገባ የምናገኘው ውጤት ነው። እንደሚከተለው መቀመር እንችላለን። 

In [26]:
# ስታንዳርድ ዲቪየሽን 
def ስታንዳርድ_ልዩነት(ቁጥር):
    #ቅናሽን 0 አድርገን እንነሳ
    ቅናሽ = 0
    አማካኝ = sum(ቁጥር)/len(ቁጥር)
    #እያንዳንዱን እይታ አማካኙን እየቀነስን ስክዌር እናድርግ እና ቅናሽ ላይ እንደምረው 
    for እሴት in ቁጥር:
        ቅናሽ += (እሴት - አማካኝ)**(2)
    # ልዩነቱን በቁጥሩ ብዛት -1 አካፍሎ መመለስ 
    return (ቅናሽ/(len(ቁጥር)-1))**(0.5)

In [27]:
# የዲፕሎማ ስታንዳርድ ልዩነትን ለማግኘት 
ስታንዳርድ_ልዩነት(ዲፕሎማ_ተመራቂ)

697.7595195806363

ስታንዳርድ ዲቪየሽኑ ወይም ልዩነት 698 አካባቢ ነው። ይኽ ማለት የውሂባችን ስብጥር መደበኛ ወይም ኖርማል የምንለው ከሆነ 68% (የውሂባችን 68 እጁ) ከአማካኙ አንድ ስታንዳርድ ዲቪየሽን ይርቃል ማለት ነው። አማካኛችን 771 ስለሆነ የዳታችን እሴት 68 በመቶው የሚገኘው 771$\pm$698 ሲገኝ የውሂባችን 95 እጁ ደግሞ 771$\pm$2(698) እንዲሁም 99 በመቶ የሚሆነው 771$\pm$3(698) ይገኛል ማለት ነው። ኢመደበኛ (ነን ኖርማል) ለሆነም እንዲሁ ስታንዳርድ ዲቪየሽንን የምንረዳበትን አግባብ [እዚህ](https://en.wikipedia.org/wiki/Chebyshev's_inequality) መመልከት ይቻላል።  

# ማጠቃለያ

መሰረታዊ የሆኑትን ገላጭ ስታትስቲክስ ለመመልከት ሞክረናል። በቀጣይ ተዛማጅነትን (ኮሪሌሽንን)  የምንገልጽባቸውን፤ የመተማመን ልዩነትን (ኮንፊደንስ ኢንተርቫል) እና ያገኘነው ውጤት ምን ያህል የተረጋገጠ ውሳኔ ለመስጠት እንደሚረዳን ለመግለጽ የምንጠቀምባቸውን የማረጋገጫ ፈተናዎች (አጋማሽን በመጠቀም የሚቀመሩ አሉ፣ አማካኝን በማነጻጸር የሚሰሩ አሉ)፤ እነዚህን በቀጣይነት እንመለከታለን። 